src/execution_plan/optimizations/apply_join.c

```diff
@@ -182,38 +129,24 @@ static void _reduce_cp_to_hashjoin(ExecutionPlan *plan, OpBase *cp) {
 			// The entire Cartesian Product can be replaced with the join op.
 			ExecutionPlan_ReplaceOp(plan, cp, value_hash_join);
 			OpBase_Free(cp);
-			pending_filters_lookup_root = value_hash_join;
-			break;
+			// Try to popagate up the remaining filters.
+			i++;
+			for(; i < filter_count; i++) {
+				OptimizeUtils_MigrateFilterOp(plan, value_hash_join, filter_ops[i]);
+			}
 		} else {
 			// The Cartesian Product still has a child operation; introduce the join op as another child.
 			ExecutionPlan_AddOp(cp, value_hash_join);
 			// Streams are no longer valid since cartesian product changed.
 			for(int j = 0; j < stream_count; j ++) raxFree(stream_entities[j]);
```

Comment:
Can break here and avoid redundant `OptimizeUtils_BuildStreamFromOp` calls `if (i+1 == filter_count)`.

Reply:
done

Reply:
The raxFree calls should be in the conditional block, as otherwise we'll double free in cleanup.